✨ Add PyTorch image classification example

nateraw opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Adds PyTorch image classification example. For now, it uses `torchvision.datasets.ImageFolder` to load local image folders (just like the flax image classification example). In the future, we will switch to using the `datasets` package's image folder (once it exists). 

Marking as draft for now as I'm still working through cleaning up changes I made from [this example](https://github.com/nateraw/vision/tree/main/image-classification) I wrote earlier that uses `datasets` instead.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

nateraw committed:
:sparkles: add pytorch image classification example

nateraw committed:
:fire: remove utils.py

nateraw committed:
:lipstick: fix flake8 style issues

nateraw committed:
:fire: remove unnecessary line

nateraw committed:
:sparkles: limit dataset sizes

NielsRogge commented:
Nice!! Relevant for #13080

sgugger commented on examples/pytorch/vision/run_image_classification.py:
```diff
@@ -0,0 +1,313 @@
+            )
+
+    normalize = T.Normalize(mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5])
+    train_dataset = ImageFolder(
```
It would be nice to use the datasets library here, or support both use cases (like run_glue for text classification).

nateraw replied:
Using `datasets` now. For image folder, we can use [`nateraw/image-folder`](https://huggingface.co/datasets/nateraw/image-folder/) until the [official implementation](https://github.com/huggingface/datasets/pull/2830) gets merged.

sgugger commented on examples/pytorch/vision/run_image_classification.py:
```diff
@@ -0,0 +1,313 @@
+        id2label[str(i)] = label
+
+    if data_args.validation_dir is not None:
+        eval_dataset = ImageFolder(
+            data_args.validation_dir,
+            T.Compose(
+                [
+                    T.Resize(data_args.image_size),
+                    T.CenterCrop(data_args.image_size),
+                    T.ToTensor(),
+                    normalize,
+                ]
+            ),
+        )
```
Let's define the validation transform in another line as black makes this hard to read.

sgugger commented on examples/pytorch/vision/run_image_classification.py:
```diff
@@ -0,0 +1,313 @@
+    normalize = T.Normalize(mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5])
+    train_dataset = ImageFolder(
+        data_args.train_dir,
+        T.Compose(
+            [
+                T.RandomResizedCrop(data_args.image_size),
+                T.RandomHorizontalFlip(),
+                T.ToTensor(),
+                normalize,
+            ]
+        ),
```
Let's define the training transform on another line as black makes this hard to read.

sgugger commented on examples/pytorch/vision/run_image_classification.py:
```diff
@@ -0,0 +1,313 @@
+        train_dataset = torch.utils.data.Subset(train_dataset, indices[:-n_val])
+        eval_dataset = torch.utils.data.Subset(train_dataset, indices[-n_val:])
+
+    test_dataset = (
+        ImageFolder(
+            data_args.test_dir,
+            T.Compose(
+                [
+                    T.Resize(data_args.image_size),
+                    T.CenterCrop(data_args.image_size),
+                    T.ToTensor(),
+                    normalize,
+                ]
+            ),
+        )
+        if data_args.test_dir is not None
+        else None
+    )
```
Let's reuse the validation transform from before and do an if then else on several lines. There is no point using the one-line syntax `foo = bar if xxx else yyy` when black splits it in several lines.

sgugger commented on examples/pytorch/vision/run_image_classification.py:
```diff
@@ -0,0 +1,313 @@
+        indices = torch.randperm(data_args.max_predict_samples).tolist()
+        test_dataset = torch.utils.data.Subset(test_dataset, indices)
+
+    # NOTE - We aren't directly using this feature extractor since we defined custom transforms above.
```
Explain a bit more why we are defining it anyway? It doesn't make sense otherwise.

sgugger reviewed (approved):
Thanks a lot for adding this! I left a few comments.

nateraw committed:
:pushpin: update reqs

nateraw committed:
:art: restructure - use datasets lib

nateraw commented on examples/pytorch/vision/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+-f https://download.pytorch.org/whl/torch_stable.html
+torch==1.9.0+cpu 
+-f https://download.pytorch.org/whl/torch_stable.html
+torchvision==0.10.0+cpu
```
resolved!

nateraw committed:
:art: import transforms directly

nateraw commented on examples/pytorch/vision/run_image_classification.py:
```diff
@@ -0,0 +1,313 @@
+import datasets
+import numpy as np
+import torch
+import torchvision.transforms as T
```
importing them directly now

NielsRogge commented:
I'll review this PR in detail (thanks for working on this!). Regarding the fixtures for the tests, I've recently moved these files to the hf-internal-testing organization on the [hub](https://huggingface.co/hf-internal-testing). This makes it more clear, as otherwise these fixture files are also downloaded when people do a `git clone` of the library.

nateraw committed:
:memo: add comments

nateraw committed:
:lipstick: style

nateraw committed:
:fire: remove flag

nateraw committed:
:pushpin: update requirement warning

nateraw committed:
:memo: add vision README.md

nateraw committed:
:memo: update README.md

nateraw committed:
:memo: update README.md

nateraw committed:
:art: add image-classification tag to model card

NielsRogge commented on examples/pytorch/test_examples.py:
```diff
@@ -340,3 +342,35 @@ def test_run_translation(self):
+            run_image_classification.py
+            --output_dir {tmp_dir}
+            --model_name_or_path google/vit-base-patch16-224-in21k
+            --train_dir tests/fixtures/tests_samples/cats_and_dogs/
```
Can we add the cats and dogs as a dataset to the hub under the [hf-internal-testing](https://huggingface.co/hf-internal-testing) organization?

nateraw replied:
We can, but this is actually testing the fact that the script works on local image folders. I can issue another PR that directly pulls them down into an imagefolder-like cache dir to test. But I'll leave this as-is for now if its not a big deal.

NielsRogge commented on examples/pytorch/vision/README.md:
```diff
@@ -0,0 +1,131 @@
+
+# Image classification examples
+
+The following example showcases how to fine-tune a `ViT` for image-classification using PyTorch.
```
```suggestion
The following examples showcase how to fine-tune a `ViT` for image-classification using PyTorch.
```

NielsRogge commented on examples/pytorch/vision/README.md:
````diff
@@ -0,0 +1,131 @@
+
+0. If you haven't already, [sign up](https://huggingface.co/join) for a 🤗 account
+
+1. Make sure you have `git-lfs` installed.
+
+```bash
+$ apt install git-lfs
+```
````
```suggestion
1. Make sure you have `git-lfs` installed and git setup.

```bash
$ apt install git-lfs
git config --global user.email "you@example.com"
git config --global user.name "Your Name"
```
```
Not sure if this is needed, but if you run the script in Google Colab for example then you also need to setup git (based on Sylvain's notebooks such as [this one](https://github.com/huggingface/notebooks/blob/master/examples/text_classification.ipynb)).

NielsRogge commented on examples/pytorch/vision/README.md:
````diff
@@ -0,0 +1,131 @@
+$ apt install git-lfs
+```
+
+2. Log in using `huggingface-cli`
````
```suggestion
2. Log in with your HuggingFace account credentials using `huggingface-cli`
```

NielsRogge commented on examples/pytorch/vision/README.md:
````diff
@@ -0,0 +1,131 @@
+
+3. When running the script, pass the following arguments:
+
+```bash
+python run_image_classification.py \
+    --push_to_hub \
+    --push_to_hub_model_id <name-your-model> \
````
Is it also possible to push to a certain organization?

nateraw replied:
Yup, this is all handled through trainer arguments

NielsRogge commented on examples/pytorch/vision/run_image_classification.py:
```diff
@@ -0,0 +1,360 @@
+from transformers.utils.versions import require_version
+
+
+""" Finetuning a 🤗 Transformers model for image classification"""
```
```suggestion
""" Fine-tuning a 🤗 Transformers model for image classification"""
```

NielsRogge commented on examples/pytorch/vision/run_image_classification.py:
```diff
@@ -0,0 +1,360 @@
+class DataTrainingArguments:
+    """
+    Arguments pertaining to what data we are going to input our model for training and eval.
+    Using `HfArgumentParser` we can turn this class
+    into argparse arguments to be able to specify them on
+    the command line.
```
```suggestion
    Using `HfArgumentParser` we can turn this class into argparse arguments to be able to 
    specify them on the command line.
```

NielsRogge commented on examples/pytorch/vision/run_image_classification.py:
```diff
@@ -0,0 +1,360 @@
+            "value if set."
+        },
+    )
+    image_size: Optional[int] = field(default=224, metadata={"help": " The size (resolution) of each image."})
```
Will images be squared? Or is it just the smaller edge of the image that will be matched to this number?

nateraw replied:
You're correct, this value is passed directly to torchvision's `Resize`

NielsRogge commented on examples/pytorch/vision/run_image_classification.py:
```diff
@@ -0,0 +1,360 @@
+    dataset_config_name: Optional[str] = field(
+        default=None, metadata={"help": "The configuration name of the dataset to use (via the datasets library)."}
+    )
+    train_dir: Optional[str] = field(default=None, metadata={"help": "A folder containing the training data."})
+    validation_dir: Optional[str] = field(default=None, metadata={"help": "A folder containing the validation data."})
+    train_val_split: Optional[float] = field(
+        default=0.15, metadata={"help": "Percent to split off of train for validation."}
+    )
```
Hmm so you have a training dataset, a validation dataset and a test set. The `validation_dir` is actually the test set? And the `train_val_split` will split the training into training + validation?

nateraw replied:
there's no test set here. There's just train + validation. If 'validation' key is not found in dataset, we create a split off of train for validation and set it to the 'validation' key of the dataset dict.

NielsRogge reviewed (commented):
Overall LGTM! Just some small comments.

Most importantly, it would be great to remove the fixtures files from this PR in favor of a HuggingFace dataset.

sgugger commented:
Last nit on my side: can we move the vision folder to be `image-classification`? We will have other kinds of vision examples in the future.

patrickvonplaten commented on examples/pytorch/vision/README.md:
```diff
@@ -0,0 +1,131 @@
+<!---
```
@sgugger - is `pytorch/vision` not a bit too general? Given that we have `examples/pytorch/text-classification/run_glue.py` I think the better structure would be `examples/pytorch/image-classification/run_image_classification.py` here no?

patrickvonplaten replied:
Never mind - just saw your comment!

nateraw committed:
:truck: rename vision ➡️ image-classification

nateraw committed:
:memo: update image-classification README.md

nateraw commented:
Ok, addressed most of the comments. Merging as-is for now.

@NielsRogge I did not address these two items, however I can in future PRs (if need be):

- Adding test data to `datasets` library. 
- Adjusting train/validation/test split logic.

Hecim1984 commented:
13134
